src/main/java/bio/terra/app/controller/DuosApiController.java

```diff
@@ -0,0 +1,83 @@
+package bio.terra.app.controller;
+
+import bio.terra.app.configuration.ApplicationConfiguration;
+import bio.terra.common.iam.AuthenticatedUserRequest;
+import bio.terra.common.iam.AuthenticatedUserRequestFactory;
+import bio.terra.controller.DuosApi;
+import bio.terra.model.DuosFirecloudGroupModel;
+import bio.terra.service.auth.iam.IamAction;
+import bio.terra.service.auth.iam.IamResourceType;
+import bio.terra.service.auth.iam.IamService;
+import bio.terra.service.duos.DuosService;
+import com.fasterxml.jackson.databind.ObjectMapper;
+import io.swagger.annotations.Api;
+import java.util.List;
+import java.util.Optional;
+import javax.servlet.http.HttpServletRequest;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import org.springframework.http.ResponseEntity;
+import org.springframework.stereotype.Controller;
+
+@Controller
+@Api(tags = {"duos"})
+public class DuosApiController implements DuosApi {
+
+  private final Logger logger = LoggerFactory.getLogger(DuosApiController.class);
+
+  private final HttpServletRequest request;
+  private final ObjectMapper objectMapper;
+  private final AuthenticatedUserRequestFactory authenticatedUserRequestFactory;
+  private final ApplicationConfiguration appConfig;
+  private final IamService iamService;
+  private final DuosService duosService;
+
+  public DuosApiController(
+      HttpServletRequest request,
+      ObjectMapper objectMapper,
+      AuthenticatedUserRequestFactory authenticatedUserRequestFactory,
+      ApplicationConfiguration appConfig,
+      IamService iamService,
+      DuosService duosService) {
+    this.request = request;
+    this.objectMapper = objectMapper;
+    this.authenticatedUserRequestFactory = authenticatedUserRequestFactory;
+    this.appConfig = appConfig;
+    this.iamService = iamService;
+    this.duosService = duosService;
+  }
+
+  @Override
+  public ResponseEntity<List<DuosFirecloudGroupModel>> retrieveDuosFirecloudGroups() {
+    return ResponseEntity.ok(duosService.retrieveFirecloudGroups());
+  }
+
+  @Override
+  public ResponseEntity<DuosFirecloudGroupModel> retrieveDuosFirecloudGroup(String duosId) {
```

Comment:
Is the underlying DuosID a UUID? If so, what would you think about switching to passing around the UUID instead of the string?

Historically, we got into a weird state with snapshot & dataset ids where we were having to switch back and forth between string & UUID throughout the code. It's my opinion we should pass around the UUID and only convert to a string directly where used in string format.

However, I could see this being a special case where passing the string makes sense because we never actually use the duosID in our code and only pass it to DUOS (in string format).

And apologies, I think this is baked into some of your previous changes, I should I have made this comment earlier in case we do want to change it.

Reply:
I am in complete agreement with you, re: UUID conversion.  In this case, duosId is not a UUID - it's a true string of the format `DUOS-<nnnnnn>` like `DUOS-123456`.  I give example values in the OpenAPI spec but I don't think I do in the code itself.

Would it help if I updated these parameter descriptions to clarify the format?  My concern with that is that we don't enforce this format ourselves: we delegate it to DUOS, so that if they change their formatting requirements we're not out of sync.

Reply:
Ah, sorry I missed that - The example in the open API spec is sufficient.